server/liveblog/themes/themes.py

```diff
@@ -77,43 +96,127 @@ class ThemesResource(Resource):
             else:
-                self.create([theme])
-                created.append(theme)
-        return (created, updated)
+                return dict(status='unchanged', theme=theme)
+        else:
+            self.create([theme])
+            return dict(status='created', theme=theme)
 
     def on_delete(self, deleted_theme):
         global_default_theme = get_resource_service('global_preferences').get_global_prefs()['theme']
         # raise an exception if the removed theme is the default one
         if deleted_theme['name'] == global_default_theme:
-            raise SuperdeskApiError.forbiddenError("This is a default theme and can not be deleted")
+            raise SuperdeskApiError.forbiddenError('This is a default theme and can not be deleted')
+        # raise an exception if the removed theme has children
+        if self.get(req=None, lookup={'extends': deleted_theme['name']}).count() > 0:
+            raise SuperdeskApiError.forbiddenError('This theme has children. It can\'t be removed')
         # update all the blogs using the removed theme and assign the default theme
         blogs_service = get_resource_service('blogs')
+        default_theme = blogs_service.get_theme_snapshot(global_default_theme)
         blogs = blogs_service.get(req=None, lookup={'theme._id': deleted_theme['_id']})
         for blog in blogs:
             # will assign the default theme to this blog
-            default_theme = blogs_service.get_theme_snapshot(global_default_theme)
             blogs_service.system_update(ObjectId(blog['_id']), {'theme': default_theme}, blog)
 
+    def get_dependencies(self, theme_name, deps=[]):
+        ''' return a list of the dependencies names '''
+        deps.append(theme_name)
+        theme = self.find_one(req=None, name=theme_name)
+        if not theme:
+            raise SuperdeskError(400, 'Dependencies are not satisfied')
+        if theme and theme.get('extends', False):
+            self.get_dependencies(theme.get('extends'), deps)
+        return deps
+
+
+@upload_theme_blueprint.route('/theme-upload', methods=['POST'])
+@cross_origin()
+def upload_a_theme():
+    themes_service = get_resource_service('themes')
+    with zipfile.ZipFile(request.files['media']) as zip_file:
+        # Keep only actual files (not folders)
+        files = [file for file in zip_file.namelist() if not file.endswith('/')]
+        # Determine if there is a root folder
+        roots = set(file.split('/')[0] for file in files)
+        root_folder = len(roots) == 1 and '%s/' % roots.pop() or ''
+        # Check if the package is correct
+        try:
+            description_file = next((file for file in files if file.endswith('theme.json')))
+            # decode and load as a json file
+            description_file = json.loads(zip_file.read(description_file).decode('utf-8'))
+        except StopIteration:
+            return json.dumps(dict(error='A theme needs a theme.json file')), 400
```

Comment:
Instead of catching the `StopIteration` error, you can also simply pass a default value argument to `next()`:

```py
item = next(<iterable>, None)
if item is None:
    # return error

# do something with item...
```